Sources/MlemMiddleware/API Client/ApiClient+Post.swift

```diff
@@ -68,15 +68,94 @@ extension ApiClient: PostFeedProvider {
         return nil
     }
     
+    /// Mark the given post as read. Works on all versions.
+    /// On v0.19.0 and above, calling this will also mark any queued posts as read unless `includeQueuedPosts` is set to `false`.
+    public func markPostAsRead(
+        id: Int,
+        read: Bool = true,
+        includeQueuedPosts: Bool = true,
+        semaphore: UInt? = nil
+    ) async throws {
+        // We *must* use `postId` in 0.18 versions, and we *must* use `postIds` from 0.19.4 onwards.
+        let version = try await version
+        let request: MarkPostAsReadRequest
+        if version >= .v19_0 {
+            try await self.markPostsAsRead(
+                ids: [id],
+                read: read,
+                includeQueuedPosts: includeQueuedPosts,
+                semaphore: semaphore
+            )
+        } else {
+            request = MarkPostAsReadRequest(postId: id, read: read, postIds: nil)
+            let response = try await perform(request)
+            if !response.success {
+                throw ApiClientError.unsuccessful
+            }
+            markReadQueue.remove(id)
+            if let post = caches.post2.retrieveModel(cacheId: id) {
+                post.readManager.updateWithReceivedValue(read, semaphore: semaphore)
+                post.readQueued = false
+            }
+        }
+    }
+    
+    /// Mark the given posts as read. Only works on v0.19.0 and above; on lower versions, use `markPostAsRead` instead.
+    /// Calling this will also mark any queued posts as read unless `includeQueuedPosts` is set to `false`.
+    public func markPostsAsRead(
+        ids: Set<Int>,
+        read: Bool = true,
+        includeQueuedPosts: Bool = true,
+        semaphore: UInt? = nil
+    ) async throws {
+        let version = try await version
+        guard version >= .v19_0 else { throw ApiClientError.unsupportedLemmyVersion }
+        
+        let idsToSend: Set<Int>
+        let markReadQueueCopy: Set<Int>
+        if read, includeQueuedPosts {
+            idsToSend = ids.union(self.markReadQueue)
+            markReadQueueCopy = self.markReadQueue
+            self.markReadQueue.removeAll()
+        } else {
+            markReadQueueCopy = []
+            idsToSend = ids
+        }
+        
+        guard !idsToSend.isEmpty else { return }
+        
+        do {
+            let request = MarkPostAsReadRequest(postId: nil, read: read, postIds: Array(idsToSend))
+            let response = try await perform(request)
+            if !response.success {
+                throw ApiClientError.unsuccessful
+            }
+            markReadQueue.subtract(ids)
```

Comment:
Why is this line necessary? Appears duplicate with L119

Reply:
When `includeQueuedPosts` is `false`, line 119 won't run. This line isn't strictly necessary, but will remove the IDs from the queue if they are present, since we just marked them as read.